src/data/set/Union_lift.lean

```diff
@@ -0,0 +1,235 @@
+  `lift_of_eq_Union_unary` -/
+lemma Union_lift_unary
+  {S : ι → set α}
+  {f : Π i (x : S i), β}
+  {hf : ∀ i j (x : α) (hxi : x ∈ S i) (hxj : x ∈ S j), f i ⟨x, hxi⟩ = f j ⟨x, hxj⟩}
+  (u : (Union S) → (Union S))
+  (ui : Π i, S i → S i)
+  (hui : ∀ i x, set.inclusion (set.subset_Union S i) (ui i x) =
+    u (set.inclusion (set.subset_Union S i) x))
+  (uβ : β → β)
+  (h : ∀ i (x : S i), (f i (ui i x)) = uβ (f i x))
+  (x : Union S) :
+  Union_lift S f hf (u x) = uβ (Union_lift S f hf x) :=
+begin
+  cases set.mem_Union.1 x.prop with i hi,
+  rw [Union_lift_of_mem x hi, ← h i],
+  have : x = (set.inclusion (set.subset_Union S i) ⟨x, hi⟩), { cases x, refl },
+  have hx' : (set.inclusion (set.subset_Union S i) (ui i ⟨x, hi⟩) : α) ∈ S i,
+    from  (ui i ⟨x, hi⟩).prop,
+  conv_lhs { rw [this, ← hui, Union_lift_of_mem _ hx'] },
+  simp only [coe_inclusion, subtype.coe_eta]
+end
+
+/-- `Union_lift_binary` is useful for proving that `Union_lift` is a homomorphism
+  of algebraic structures when defined on the Union of algebraic subobjects.
+  For example, it could be used to prove that the lift of a collection
+  of group homomorphisms on a union of subgroups preserves `*`. See also
+  `lift_of_eq_Union_binary` -/
+lemma Union_lift_binary
+  {S : ι → set α}
+  {f : Π i (x : S i), β}
+  {hf : ∀ i j (x : α) (hxi : x ∈ S i) (hxj : x ∈ S j), f i ⟨x, hxi⟩ = f j ⟨x, hxj⟩}
+  (dir: directed (≤) S)
+  (op : (Union S) → (Union S) → (Union S))
+  (opi : Π i, S i → S i → S i)
+  (hopi : ∀ i x y, set.inclusion (set.subset_Union S i) (opi i x y) =
+    op (set.inclusion (set.subset_Union S i) x)
+       (set.inclusion (set.subset_Union S i) y))
+  (opβ : β → β → β)
+  (h : ∀ i (x y : S i), (f i (opi i x y)) = opβ (f i x) (f i y))
+  (x y : Union S) :
+  Union_lift S f hf (op x y) = opβ (Union_lift S f hf x) (Union_lift S f hf y) :=
+begin
+  cases set.mem_Union.1 x.prop with i hi,
+  cases set.mem_Union.1 y.prop with j hj,
+  rcases dir i j with ⟨k, hik, hjk⟩,
+  rw [Union_lift_of_mem x (hik hi), Union_lift_of_mem y (hjk hj), ← h k],
+  have hx : x = (set.inclusion (set.subset_Union S k) ⟨x, hik hi⟩), { cases x, refl },
+  have hy : y = (set.inclusion (set.subset_Union S k) ⟨y, hjk hj⟩), { cases y, refl },
+  have hxy : (set.inclusion (set.subset_Union S k) (opi k ⟨x, hik hi⟩ ⟨y, hjk hj⟩) : α) ∈ S k,
+    from (opi k ⟨x, hik hi⟩ ⟨y, hjk hj⟩).prop,
+  conv_lhs { rw [hx, hy, ← hopi, Union_lift_of_mem _ hxy] },
+  simp only [coe_inclusion, subtype.coe_eta]
+end
+
+/-- Version of `Union_lift` for sets that are propositionally equal to the Union.
+  Define a map on a set equal to the `Union` of sets by defining it on each of the sets
+  in the `Union`. -/
+noncomputable def lift_of_eq_Union
+  (S : ι → set α)
+  (f : Π i (x : S i), β)
+  (hf : ∀ i j (x : α) (hxi : x ∈ S i) (hxj : x ∈ S j), f i ⟨x, hxi⟩ = f j ⟨x, hxj⟩)
+  (T : set α) (hT : T = ⋃ i, S i)
+  (x : T) : β :=
+by subst hT; exact Union_lift S f hf x
```

Comment:
I'm not sure I understand the point of this def; can't the caller just do the `subst`?

Reply:
The `subst` doesn't work when `T` is a more complicated expression.